### ISSUE TITLE
Typechecking regression between OCaml 4.12 and OCaml 4.13.x?

### ISSUE THREAD

**Issue by francoisthire** [ocaml/ocaml#10768]
I have a snippet that type checks with OCaml 4.12.0 but does not with OCaml 4.13.0 nor OCaml 4.13.1. I was wondering whether it was a regression

Here is the original code which type checks with OCaml 4.12.0:

```ocaml
type _ t

module type S = sig type t end

type pack = Pack : 'a t *  (module S with type t = 'a) -> pack

let dispatch : pack list ref = ref []

exception Not_registered

type (_,_) equal =
  | Refl : 'a -> ('a,'a) equal

type equality =
  {
    equal: 'a 'b. ('a t * 'b t) -> ('a t,'b t) equal option
  }

let equal : equality ref = ref {equal = fun _ -> None}

let get (type a) : a t -> (module S with type t = a)  = fun index ->
  let dispatch = !dispatch in
  let rec unpack list =
    match list with
    | [] -> raise Not_registered
    | Pack (index', (module P))::list' ->
      match !equal.equal (index, index') with
      | Some (Refl _) ->  (module P : ( S with type t = a))
        | None -> unpack list'
  in
  unpack dispatch
```
With OCaml 4.13.1 the code fails with the following error:

```
File "test.ml", line 28, characters 34-35:
28 |       | Some (Refl _) ->  (module P : ( S with type t = a))
                                       ^
Error: Signature mismatch:
       Modules do not match:
         sig type t = $Pack_'a end
       is not included in
         sig type t = P.t end
       Type declarations do not match:
         type t = $Pack_'a
       is not included in
         type t = P.t
       File "test.ml", line 3, characters 20-26: Expected declaration
       File "test.ml", line 3, characters 20-26: Actual declaration
```

I managed to write fix that type checks with OCaml 4.12, 4.13 and 4.13.1 for the get function:

```ocaml
let get (type a) : a t -> (module S with type t = a)  = fun index ->
  let dispatch = !dispatch in
  let rec unpack list : (module S with type t = a) =
    match list with
    | [] -> raise Not_registered
    | Pack (index', (module P : (S with type t = _)))::list' ->
      match !equal.equal (index, index') with
      | Some (Refl _) ->  (module (struct include P end) : (S with type t = _))
      | None -> unpack list'
  in
  unpack dispatch
```

There are two differences: a type annotation for the function unpack and module P is replaced by module (struct include P end). I was not able to write a function that type checks for OCaml 4.13.x without the latter change.

I have read the Changelog of OCaml 4.13, but I was not able to figure out whether this is a regression.

**Comment by lthls** [comment]
I'm not an expert on the subject, but I think that you're unpacking your first-class module too early (i.e. before matching on the equality).
Here is a version that compiles on 4.13, which doesn't unpack the module at all:
```ocaml
let get (type a) : a t -> (module S with type t = a)  = fun index ->
  let dispatch = !dispatch in
  let rec unpack list : (module S with type t = a) =
    match list with
    | [] -> raise Not_registered
    | Pack (index', m)::list' ->
      match !equal.equal (index, index') with
      | Some (Refl _) -> m
      | None -> unpack list'
  in
  unpack dispatch
```
If you actually need to use functions or types from the module, then I've checked that it also works if you unpack in the `Refl` branch:
```ocaml
      | Some (Refl _) ->
        let module P = (val m : S with type t = a) in
        (module P)
```
You can also move the type annotation that I've put on the `unpack` function back on the returned value if you prefer.

What I can't tell, however, is whether your original example should work with 4.13 (showing a regression), shouldn't work in 4.12 (showing a bug that got fixed), or is in a grey zone where type inference is known to be undecidable. Hopefully someone with more experience could clarify this.

**Comment by francoisthire** [comment]
Just to be sure:

```ocaml
      | Some (Refl _) ->
        let module P = (val m : S with type t = a) in
        (module P)
```

should be

```ocaml
      | Some (Refl _) ->
        let module P = (module (val m : S with type t = a)) in
        (module P)
```

Which is quite similar to the `module (struct  include P end)` I had in the first message. 

However, your initial example is indeed valid. Which makes me wonder why? There is a subtlety that I don't get. Why does unpacking too soon would make the type checking fails?

**Comment by lthls** [comment]
I believe my suggestion is the correct one. `let module M = ...` binds a normal module, so the right-hand side must be a module expression. It might be counter-intuitive, but `(module M)` is a "simple" expression representing a first-class module, while `(val m)` is a module expression.
But you could also use `let (module P : S with type t = a) = m in (module P)`.
(The difference is that `let module P` binds a "second-class" module to a module expression, like in `let module ISet = Set.Make(Int)`, while `let (module P)` expects an expression for a first-class module and then unpacks it, binding the result to `P`, like in `let (module ISet) = make_set Int.compare`)

Regarding the initial type error, if my understanding is correct, binding `P` in the pattern matching for `Pack` will make `P` have type `sig type t = $Pack_'a end`, which is not the same as `sig type t = a end` at this point. Then in the branch, I assume that in 4.12 `(module P)` is interpreted as a new module, whose implementation happens to be based on `P`, but which can have a different module type, while in 4.13 `(module P)` needs to have the same module type as `P`, so you need to explicitly create a new module using `struct include P end` to get the type you want.
By unpacking the module in the branch, you can give your module the right type directly and avoid the issue (the types `$Pack_'a` and `a` are the same whenever `P` is in scope)

**Comment by mrmr1993** [comment]
Minimal example derived from some code I'm working on:
```ocaml
type (_, _) refl = Refl : ('a, 'a) refl;;

module type S = sig
  type t
end;;

module type S' = sig
  type _ t_aux

  type t

  val eq : (t, unit t_aux) refl
end;;

let f (type t) (module M : S' with type t = t) : (module S with type t = t) =
  let Refl = M.eq in
  (module M);;
```
Which gives the error
```
Error: Signature mismatch:
       Modules do not match:
         sig
           type 'a t_aux = 'a M.t_aux
           type t = M.t
           val eq : (t, unit t_aux) refl
         end
       is not included in
         sig type t = M.t end
       Type declarations do not match:
         type t = M.t
       is not included in
         type t = M.t
       The type t = unit t_aux is not equal to the type t = unit t_aux
```

Planning to look into this this evening.

**Comment by mrmr1993** [comment]
Since this typechecks:
```ocaml
let f (type t) (module M : S' with type t = t) : (module S with type t = t) =
  let Refl = M.eq in
  let module N : S' with type t = t = M in
  (module N);;
```
it's likely a spurious scope escape due to the expansion of the newtype.

**Comment by mrmr1993** [comment]
A bisect points to eed1110e6aece8713605806bee7066ea07e2c0ca, coming from PR #10277.

Indeed, this issue was [already anticipated there](https://github.com/ocaml/ocaml/pull/10277#issuecomment-819446345).

**Comment by mrmr1993** [comment]
Proposed fix: #11339.
